fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/CreateResourceServiceImpl.java

```diff
@@ -0,0 +1,229 @@
+
+import static org.fcrepo.kernel.api.FedoraTypes.FEDORA_CREATED;
+import static org.fcrepo.kernel.api.FedoraTypes.FEDORA_PAIRTREE;
+import static org.fcrepo.kernel.api.RdfLexicon.NON_RDF_SOURCE;
+import static org.fcrepo.kernel.api.rdf.DefaultRdfStream.fromModel;
+import static org.fcrepo.kernel.impl.services.functions.FedoraUtils.addToIdentifier;
+
+import org.apache.jena.datatypes.xsd.XSDDatatype;
+import org.apache.jena.graph.Triple;
+import org.apache.jena.rdf.model.Model;
+import org.fcrepo.kernel.api.models.ExternalContent;
+import org.fcrepo.kernel.api.RdfStream;
+import org.fcrepo.kernel.api.exception.CannotCreateResourceException;
+import org.fcrepo.kernel.api.exception.ItemNotFoundException;
+import org.fcrepo.kernel.api.exception.RepositoryRuntimeException;
+import org.fcrepo.kernel.api.models.ResourceHeaders;
+import org.fcrepo.kernel.api.operations.NonRdfSourceOperationBuilder;
+import org.fcrepo.kernel.api.operations.NonRdfSourceOperationFactory;
+import org.fcrepo.kernel.api.operations.RdfSourceOperationFactory;
+import org.fcrepo.kernel.api.operations.ResourceOperation;
+import org.fcrepo.kernel.api.services.CreateResourceService;
+import org.fcrepo.kernel.api.services.functions.UniqueValueSupplier;
+import org.fcrepo.kernel.impl.operations.AbstractResourceOperation;
+import org.fcrepo.persistence.api.PersistentStorageSession;
+import org.fcrepo.persistence.api.PersistentStorageSessionManager;
+import org.fcrepo.persistence.api.exceptions.PersistentItemNotFoundException;
+import org.fcrepo.persistence.api.exceptions.PersistentStorageException;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.Link;
+
+import java.io.InputStream;
+import java.net.URI;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Create a RdfSource resource.
+ * @author whikloj
+ * TODO: bbpennel has thoughts about moving this to HTTP layer.
+ */
+public class CreateResourceServiceImpl extends AbstractService implements CreateResourceService {
+
+    @Inject
+    private PersistentStorageSessionManager psManager;
+
+    @Inject
+    private RdfSourceOperationFactory rdfSourceOperationFactory;
+
+    @Inject
+    private NonRdfSourceOperationFactory nonRdfSourceOperationFactory;
+
+    @Inject
+    private UniqueValueSupplier minter;
+
+    @Override
+    public void perform(final String txId, final String fedoraId, final String slug, final boolean isContained,
+                        final String contentType, final List<String> linkHeaders, final Collection<String> digest,
+                        final InputStream requestBody, final ExternalContent externalContent) {
+        final PersistentStorageSession pSession = this.psManager.getSession(txId);
+        checkAclLinkHeader(linkHeaders);
+        checkParent(pSession, (isContained ? fedoraId : findExistingAncestor(fedoraId)));
```

Comment:
could you add some short comments explaining the isContained behaviors? There's a lot happening on these two lines, I was having a somedifficulty following the cases and the expected outcomes. Putting the result of this ternary into its own variable might help a little too.

Reply:
So `isContained` is a way to tell if the `fedoraId` is the parent of the eventual container or the final path. 
Basically 
```
if (httpRequestMethod.equals("POST")) {
   isContained = true;
else if (httpRequestMethod.equals("PUT")) {
  isContained = false;
}
```
Because when we get into the service we need to either treat `fedoraId` as the parent of the resource we are creating (POST, so check that its not a binary and that it does exist) or that `fedoraId` is the final path of the resource we are creating (PUT, so find the a containment parent from the path and check that it exists).

This might be changed once the containment index is resolved. But I wanted to make clear the difference between making a resource called `Slug` under `fedoraId` and making `fedoraId` (possibly under something).